src/test/java/io/jenkins/pluginhealth/scoring/PluginServiceTest.java

```diff
@@ -0,0 +1,65 @@
+package io.jenkins.pluginhealth.scoring;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.repository.PluginRepository;
+import io.jenkins.pluginhealth.scoring.service.PluginService;
+
+import static org.junit.Assert.assertEquals;
+
+import org.jetbrains.annotations.NotNull;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.util.TestPropertyValues;
+import org.springframework.context.ApplicationContextInitializer;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.test.context.ContextConfiguration;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import static org.mockito.Mockito.mock;
+
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@ContextConfiguration(initializers = PluginServiceTest.DockerPostgresDatasourceInitializer.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Testcontainers
+@RunWith(MockitoJUnitRunner.class)
+public class PluginServiceTest {
+    private final PluginRepository pluginRepository;
+    private final PluginService pluginService;
+
+    Plugin plugin = new Plugin("myPlugin", "https://github.com/jenkinsci/my-plugin", null);
+
+    public PluginServiceTest(PluginRepository pluginRepository, PluginService pluginService) {
+        this.pluginRepository = pluginRepository;
+        this.pluginService = pluginService;
+    }
+
+    @Container
+    private static final PostgreSQLContainer POSTGRE_SQL_CONTAINER = new PostgreSQLContainer("postgres:14.1")
+        .withDatabaseName("testdb")
+        .withUsername("sa")
+        .withPassword("sa");
+
+    public static class DockerPostgresDatasourceInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {
+        @Override
+        public void initialize(@NotNull ConfigurableApplicationContext applicationContext) {
+            TestPropertyValues.of(
+                "spring.datasource.url=" + POSTGRE_SQL_CONTAINER.getJdbcUrl(),
+                "spring.datasource.username=" + POSTGRE_SQL_CONTAINER.getUsername(),
+                "spring.datasource.password=" + POSTGRE_SQL_CONTAINER.getPassword()
+            ).applyTo(applicationContext);
+        }
+    }
+
+    @Test
+    public void myTest() {
+        pluginService.saveOrUpdate(plugin);
+        pluginService.saveOrUpdate(plugin);
+
+        assertEquals(1, pluginRepository.findAllByName("myPlugin").size());
```

Comment:
I created a new method in Repository that fetches all the plugins from the DB with a given name. And here I'm fetching all the plugins by the name "myPlugin" and ensuring that there's only 1 (even though we saved the same plugin object twice in the database, see lines 60 and 61 above).

Reply:
> I don't think we should have multiple plugin with the same name.

Thanks, I've added a `unique=true` constraint on the name field. But regarding this quoted text: yes, we should not have multiple plugins with the same name but the aim of this test is to catch that test case where we mistakenly added two plugin objects with the same name in the database. And we're catching that by checking the size of the list returned by the `findAll()` method and equating it with "1". Whenever you're free, can you please help me understand this?

Reply:
I added a test in `PluginServiceIT`, which is failing (on purpose) for you to see my point. See https://github.com/jenkins-infra/plugin-health-scoring/pull/17/commits/4474bef52444fb5290b5311ea6ad7206a538386c.

There is some mechanism in hibernate which saving twice the same object is triggering, so your test is not entirely validating the issue we have. 
For you to understand, please see that the problem we have is triggered when we start the application a second time, with the database already setup (and populated). The second time, when we read the `update-center.json`, new objects are created based on the content of the file. The values of the objects might be the same, but the objects are different (different addresses in memory). This is not what your test was trying to reproduce. 

It's important, when fixing a bug, to understand the context in which the bug is reproducible, and stick to that context / execution when writing the test(s), so that we are truly reproducing the problem. And therefore, solving it.

Reply:
This is about learning. I'm thrilled that you have!